incremental-font-transfer/src/patch_group.rs

```diff
@@ -0,0 +1,1253 @@
+                }
+            }
+
+            if accumulated_info.is_empty() {
+                return Err(PatchingError::EmptyPatchList);
+            }
+
+            self.font
+                .apply_glyph_keyed_patches(accumulated_info.into_iter())?
+        };
+
+        for info in self.non_invalidating_patch_iter() {
+            if let Some(status) = patch_data.get_mut(&info.uri) {
+                *status = UriStatus::Applied;
+            };
+        }
+
+        Ok(new_font)
+    }
+}
+
+/// Tracks whether a URI has already been applied to a font or not.
+#[derive(PartialEq, Eq, Debug)]
+pub enum UriStatus {
+    Applied,
+    Pending(Vec<u8>),
+}
+
+/// Tracks information related to a patch necessary to apply that patch.
+#[derive(PartialEq, Eq, Debug)]
+pub(crate) struct PatchInfo {
+    uri: String,
+    source_table: IftTableTag,
+    // TODO: details for how to mark the patch applied in the mapping table (ie. bit index to flip).
+    // TODO: Signals for heuristic patch selection:
+}
+
+impl PatchInfo {
+    pub(crate) fn tag(&self) -> &IftTableTag {
+        &self.source_table
+    }
+}
+
+impl From<PatchUri> for PatchInfo {
+    fn from(value: PatchUri) -> Self {
+        PatchInfo {
+            uri: value.uri_string(),
+            source_table: value.source_table(),
+        }
+    }
+}
+
+/// Type for a single non invalidating patch.
+#[derive(PartialEq, Eq, Debug)]
+struct NoInvalidationPatch(PatchInfo);
+
+/// Type for a single partially invalidating patch.
+#[derive(PartialEq, Eq, Debug)]
+struct PartialInvalidationPatch(PatchInfo);
+
+/// Type for a single fully invalidating patch.
+#[derive(PartialEq, Eq, Debug)]
+struct FullInvalidationPatch(PatchInfo);
+
+/// Represents a group of patches which are valid (compatible) to be applied together to
+/// an IFT font.
+#[derive(PartialEq, Eq, Debug)]
+enum CompatibleGroup {
+    Full(FullInvalidationPatch),
+    Mixed { ift: ScopedGroup, iftx: ScopedGroup },
+}
+
+/// A set of zero or more compatible patches that are derived from the same scope
+/// ("IFT " vs "IFTX")
+#[derive(PartialEq, Eq, Debug)]
+enum ScopedGroup {
+    PartialInvalidation(PartialInvalidationPatch),
+    NoInvalidation(BTreeMap<String, NoInvalidationPatch>),
+}
+
+impl ScopedGroup {
+    fn has_uris(&self) -> bool {
+        match self {
+            ScopedGroup::PartialInvalidation(PartialInvalidationPatch(_)) => true,
+            ScopedGroup::NoInvalidation(uri_map) => !uri_map.is_empty(),
+        }
+    }
+
+    fn no_invalidation_iter(&self) -> impl Iterator<Item = &PatchInfo> {
+        match self {
+            ScopedGroup::PartialInvalidation(_) => NoInvalidationPatchesIter { it: None },
+            ScopedGroup::NoInvalidation(map) => NoInvalidationPatchesIter {
+                it: Some(map.values()),
+            },
+        }
+    }
+}
+
+struct NoInvalidationPatchesIter<'a, T>
+where
+    T: Iterator<Item = &'a NoInvalidationPatch>,
```

Comment:
you can probably remove this bound, and just leave it on the impl block. (below, as well)

Reply:
Compiler says no :(